cmd/token.go

```diff
@@ -0,0 +1,192 @@
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/google/go-tpm-tools/client"
+	"github.com/google/go-tpm-tools/launcher/agent"
+	"github.com/google/go-tpm-tools/launcher/spec"
+	"github.com/google/go-tpm-tools/launcher/verifier"
+	"github.com/google/go-tpm-tools/launcher/verifier/rest"
+	"github.com/google/go-tpm/legacy/tpm2"
+	"github.com/spf13/cobra"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+)
+
+var mdsClient *metadata.Client
+
+// If hardware technology needs a variable length teenonce then please modify the flags description
+var tokenCmd = &cobra.Command{
+	Use:   "token",
+	Short: "Attest and fetch an OIDC token from Google Attestation Verification Service.",
+	Long: `Gather attestation report and send it to Google Attestation Verification Service for an OIDC token.
+The OIDC token includes claims regarding the GCE VM, which is verified by Attestation Verification Service. Note that Confidential Computing API needs to be enabled for your account to access Google Attestation Verification Service https://pantheon.corp.google.com/apis/api/confidentialcomputing.googleapis.com.
+--algo flag overrides the public key algorithm for the GCE TPM attestation key. If not provided then by default rsa is used.
+`,
+	Args: cobra.NoArgs,
+	RunE: func(*cobra.Command, []string) error {
+		rwc, err := openTpm()
+		if err != nil {
+			return err
+		}
+		defer rwc.Close()
+
+		// Metadata Server (MDS). A GCP specific client.
+		mdsClient = metadata.NewClient(nil)
+
+		ctx := namespaces.WithNamespace(context.Background(), namespaces.Default)
+		// TODO: principalFetcher is copied from go-tpm-tools/launcher/container_runner.go, to be refactored
```

Comment:
Please add an issue to follow up on the TODOs in this PR. Ideally, this can be a fast follow-up